presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ExtractSpatialJoins.java

```diff
@@ -331,9 +424,85 @@ else if (alignment < 0) {
                 newRightNode,
                 outputSymbols,
                 newFilter,
-                Optional.empty(),
-                Optional.empty(),
-                Optional.empty()));
+                leftPartitionSymbol,
+                rightPartitionSymbol,
+                kdbTree.map(KDB_TREE_CODEC::toJson)));
+    }
+
+    private static KDBTree loadKdbTree(String tableName, Session session, Metadata metadata, SplitManager splitManager, PageSourceManager pageSourceManager)
+    {
+        QualifiedObjectName name = toQualifiedObjectName(tableName, session.getCatalog().get(), session.getSchema().get());
+        TableHandle tableHandle = metadata.getTableHandle(session, name)
+                .orElseThrow(() -> new PrestoException(INVALID_SPATIAL_PARTITIONING, format("Table not found: %s", name)));
+        Map<String, ColumnHandle> columnHandles = metadata.getColumnHandles(session, tableHandle);
+        List<ColumnHandle> visibleColumnHandles = columnHandles.values().stream()
+                .filter(handle -> !metadata.getColumnMetadata(session, tableHandle, handle).isHidden())
+                .collect(toImmutableList());
+        if (visibleColumnHandles.size() != 1) {
+            throw new PrestoException(INVALID_SPATIAL_PARTITIONING, format("Expected single column for table %s, but found %s columns", name, columnHandles.size()));
+        }
+
+        ColumnHandle kdbTreeColumn = Iterables.getOnlyElement(visibleColumnHandles);
+
+        List<TableLayoutResult> layouts = metadata.getLayouts(session, tableHandle, Constraint.alwaysTrue(), Optional.of(ImmutableSet.of(kdbTreeColumn)));
+        verify(!layouts.isEmpty(), "layouts is empty for table: %s", name);
```

Comment:
What about empty partitioned table with zero partitions? Based on `PickTableLayout` empty layouts seems to be legit, and 0 rows is returned.